flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarPartitionSplitReaderTestBase.java

```diff
@@ -126,63 +196,91 @@ void pollMessageAfterTimeout(PulsarPartitionSplitReaderBase<String> splitReader)
                 "Couldn't poll message from Pulsar.");
     }
 
-    /** Create a split reader with max message 1, fetch timeout 1s. */
-    protected abstract PulsarPartitionSplitReaderBase<String> splitReader();
-
-    /** JUnit5 extension for all the TestTemplate methods in this class. */
-    public class PulsarSplitReaderInvocationContextProvider
-            implements TestTemplateInvocationContextProvider {
+    @Test
+    void consumeMessageCreatedAfterHandleSplitChangesAndFetch() {
+        PulsarPartitionSplitReaderBase<String> splitReader = splitReader();
+        String topicName = randomAlphabetic(10);
+        seekStartPositionAndHandleSplit(splitReader, topicName, 0);
+        operator().sendMessage(topicNameWithPartition(topicName, 0), STRING, randomAlphabetic(10));
+        fetchedMessages(splitReader, 1, true);
+    }
 
-        @Override
-        public boolean supportsTestTemplate(ExtensionContext context) {
-            return true;
-        }
+    @Test
+    void consumeMessageCreatedBeforeHandleSplitsChanges() {
+        PulsarPartitionSplitReaderBase<String> splitReader = splitReader();
+        String topicName = randomAlphabetic(10);
+        operator().setupTopic(topicName, STRING, () -> randomAlphabetic(10));
+        seekStartPositionAndHandleSplit(splitReader, topicName, 0);
+        fetchedMessages(splitReader, 0, true);
+    }
 
-        @Override
-        public Stream<TestTemplateInvocationContext> provideTestTemplateInvocationContexts(
-                ExtensionContext context) {
-            return Stream.of(new PulsarSplitReaderInvocationContext(splitReader()));
-        }
+    @Test
+    void consumeMessageCreatedBeforeHandleSplitsChangesAndResetToEarliestPosition() {
+        PulsarPartitionSplitReaderBase<String> splitReader = splitReader();
+        String topicName = randomAlphabetic(10);
+        operator().setupTopic(topicName, STRING, () -> randomAlphabetic(10));
+        seekStartPositionAndHandleSplit(splitReader, topicName, 0, MessageId.earliest);
+        fetchedMessages(splitReader, NUM_RECORDS_PER_PARTITION, true);
     }
 
-    /** Parameter resolver for Split Reader. */
-    public static class PulsarSplitReaderInvocationContext
-            implements TestTemplateInvocationContext {
+    @Test
+    @Disabled
+    void consumeMessageCreatedBeforeHandleSplitsChangesAndUseSecondMessageIdCursor() {
+        PulsarPartitionSplitReaderBase<String> splitReader = splitReader();
+        String topicName = randomAlphabetic(10);
+        operator().setupTopic(topicName, STRING, () -> randomAlphabetic(10));
+        seekStartPositionAndHandleSplit(splitReader, topicName, 0, new MessageIdImpl(0, 1, -1));
```

Comment:
Here the messageId, Guess it should be MessageId (0 , 1 ,0) instead of (0, 1 , -1) ?

Reply:
During test, I find out that the starting ledger id is 11 or 12 (seems nondeterministic to me), interesting.  Because the MessageId I passed in is lower than the oldest message id ,it has the same effect as earliest.

Reply:
Yes  0 is not the correct ledger id, it is non deterministic and should figure out a different way to get the 2nd element in the topic